src/lua/LuaBody.cpp

```diff
@@ -118,6 +118,55 @@ static int l_body_attr_path(lua_State *l)
 	return 1;
 }
 
+static int l_body_get_component(lua_State *l)
+{
+	Body *b = LuaObject<Body>::CheckFromLua(1);
+	std::string componentName = luaL_checkstring(l, 2);
+
+	BodyComponentDB::PoolBase *pool = BodyComponentDB::GetComponentType(componentName);
+	if (pool && pool->luaInterface) {
+		pool->luaInterface->PushToLua(b);
+		return 1;
+	}
+
+	if (componentName == "__properties") {
+		lua_pushnil(l);
+		return 1;
+	}
+
+	lua_getuservalue(l, 1);
+	lua_pushvalue(l, 2);
+	lua_rawget(l, -2);
+
+	return 1;
+}
+
+static int l_body_set_component(lua_State *l)
+{
+	Body *b = LuaObject<Body>::CheckFromLua(1);
+	std::string componentName = luaL_checkstring(l, 2);
+
+	BodyComponentDB::PoolBase *pool = BodyComponentDB::GetComponentType(componentName);
+	if (pool) {
+		return luaL_error(l, "Cannot set C++ body component '%s' to a lua value.");
+	}
+
+	if (componentName == "__properties") {
+		return luaL_error(l, "'__properties' is a reserved name for body components.");
```

Comment:
if I'm reading the code correctly, you don't put the components in the `__properties` subtable.

Reply:
So, before PropertyMap 2.0 the uservalue on a LuaObject was the properties table directly. We no longer use a Lua table to hold properties, and instead all properties are stored in a PropertyMap on the C++ side. The uservalue for a LuaObject is now used to store LuaComponent tables, with the __properties key reserved to hold a lightuserdata pointer to the PropertyMap for the object. This allows quickly querying properties without needing to do a dynamic cast, because the same index metamethod is used for everything from LuaVectors to Ships.

In sum, there is no __properties subtable; it's not even a table at all.

Reply:
oops really it's not a subtable.  But my complaint remains - you don't store components in an object named "__properties", but write that it is reserved for components. Or am I mistranslating the meaning?

Reply:
Hmm, good point - I wrote that when I was too tired for good english - it's a reserved name that lua components are not allowed to use, because otherwise they could inadvertently corrupt the property map pointer and cause difficult-to-debug downstream errors.

Reply:
Revised the error message, should be clearer now!